### PR TITLE
perf: keyword lookups in the tokenizer

### PR DESCRIPTION
**Pull request by vmg**
## Description
Happy Thursday everyone, this week we're bringing `sqlparser` performance improvements. I had a chance to sit with @frouioui and look at some of the profiles that we're now acquiring from his Are We Fast Yet (TM) work. There was nothing glaringly obvious that would provide massive optimization gains (as one would expect at this point, Vitess is quite optimized already), but for the normal request lifecycle, all of the `sqlparser` operations on the AST are always quite hot, and most importantly for our goals, CPU-bound.

Let's start squeezing some blood out of this stone: this particular PR comes from _allocation_ benchmarks: the code in our SQL Tokenizer that processes SQL keywords allocates so much memory that it's showing as a hotspot in a CPU profiler and _very clearly_ an allocation hotstop in the memory profiler.

Why does it do all these allocations? Well, right now it is copying the current token it's processing into a temporary buffer (this is the buffer that gets returned to the caller) and then it does yet another copy of the buffer to _lowercase it_ so it can be looked up in our keywords table (people following at home will surely remember that SQL keywords are case insensitive).

Let's improve this with some very classical Compiler Theory approaches: Instead of using a hash table to lookup keywords, use a perfect table (a perfect table is a minimal hash table where lookups cannot collide -- it is measurably faster than a normal hashtable, even the one built-in in the Go runtime). And since we now have a perfect hash table, we control the hashing algorithm used for lookups... So we can switch the algorithm to perform the hashing case-insensitively. This makes it so we don't have to create lowercase copies of all keywords!

```
name                     old time/op  new time/op  delta
Normalize-16             7.50µs ± 2%  7.38µs ± 1%    ~     (p=0.222 n=5+5)
ParseDjango-16           11.4µs ± 2%  11.0µs ± 2%  -3.69%  (p=0.016 n=5+5)
Parse1-16                14.5µs ± 2%  14.2µs ± 2%  -2.08%  (p=0.032 n=5+5)
Parse2-16                46.9µs ± 3%  44.5µs ± 3%  -5.07%  (p=0.008 n=5+5)
Parse2Parallel-16        9.29µs ± 4%  9.13µs ± 3%    ~     (p=0.548 n=5+5)
Parse3-16                5.82ms ± 1%  5.85ms ± 1%    ~     (p=0.222 n=5+5)
```

Results are :gucci: in the most realistic parse benchmarks. The pathological benchmarks do not regress.

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

## Impacted Areas in Vitess
Components that this PR will affect:

- [x]  Query Serving
- [ ]  VReplication
- [ ]  Cluster Management
- [ ]  Build/CI
- [ ]  VTAdmin

### CODE REVIEW

File: go/vt/sqlparser/keywords_test.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+func TestKeywordTable(t *testing.T) {
+	for _, kw := range keywords {
+		lookup, ok := keywordLookupTable.LookupString(kw.name)
+		if !ok {
+			t.Fatalf("keyword %q failed to match", kw.name)
+		}
+		if lookup != kw.id {
+			t.Fatalf("keyword %q matched to %d (expected %d)", kw.name, lookup, kw.id)
+		}
```
Comment (harshit-gangal) [inline go/vt/sqlparser/keywords_test.go:13]: nit: use `require` over `t.fatal`

File: go/vt/sqlparser/parse_test.go [hunk 2]
```diff
@@ -3091,21 +3092,25 @@ func TestSkipToEnd(t *testing.T) {
...
 
-func TestParseDjangoQueries(t *testing.T) {
-
-	file, err := os.Open("./test_queries/django_queries.txt")
+func loadQueries(t testing.TB, filename string) (queries []string) {
+	file, err := os.Open(path.Join("testdata", filename))
 	if err != nil {
-		t.Errorf(" Error: %v", err)
+		t.Fatal(err)
 	}
```
Comment (harshit-gangal) [inline go/vt/sqlparser/parse_test.go:3099]: nit: use `require.NoError(t, err)`

File: go/vt/sqlparser/parse_test.go [hunk 3]
```diff
@@ -3091,21 +3092,25 @@ func TestSkipToEnd(t *testing.T) {
...
 
-		_, err := Parse(string(scanner.Text()))
+func TestParseDjangoQueries(t *testing.T) {
+	for _, query := range loadQueries(t, "django_queries.txt") {
+		_, err := Parse(query)
 		if err != nil {
-			t.Error(scanner.Text())
-			t.Errorf(" Error: %v", err)
+			t.Errorf("failed to parse %q: %v", query, err)
 		}
```
Comment (harshit-gangal) [inline go/vt/sqlparser/parse_test.go:3114]: nit: use `require.NoError(t, err)`

File: go/vt/sqlparser/parse_test.go [hunk 4]
```diff
@@ -3142,25 +3147,35 @@ const (
...
 
+func BenchmarkParseDjango(b *testing.B) {
+	queries := loadQueries(b, "django_queries.txt")
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		_, err := Parse(queries[rand.Intn(len(queries))])
+		if err != nil {
+			b.Fatal(err)
+		}
```
Comment (harshit-gangal) [inline go/vt/sqlparser/parse_test.go:3158]: nit: use `require.NoError(t, err)`

File: go/vt/sqlparser/parse_test.go [hunk 5]
```diff
@@ -3142,25 +3147,35 @@ const (
...
+}
+
 func BenchmarkParse1(b *testing.B) {
 	sql := sql1
 	for i := 0; i < b.N; i++ {
-		ast, err := Parse(sql)
+		_, err := Parse(sql)
 		if err != nil {
 			b.Fatal(err)
 		}
```
Comment (harshit-gangal) [inline go/vt/sqlparser/parse_test.go:3168]: same as above.

File: go/vt/sqlparser/parse_test.go [hunk 6]
```diff
@@ -3142,25 +3147,35 @@ const (
...
 }
 
 func BenchmarkParse2(b *testing.B) {
 	sql := sql2
 	for i := 0; i < b.N; i++ {
-		ast, err := Parse(sql)
+		_, err := Parse(sql)
 		if err != nil {
 			b.Fatal(err)
 		}
```
Comment (harshit-gangal) [inline go/vt/sqlparser/parse_test.go:3178]: same as above.

**Review by harshit-gangal — APPROVED** [review]

**Comment by derekperkins** [comment]
I love how entertaining AND descriptive your PRs are @vmg :)

**Review by GuptaManan100 — COMMENTED** [review]
💯

**Review by GuptaManan100 — APPROVED** [review]

**Comment by vmg** [comment]
@harshit-gangal: I've added `testify` to all the new test cases, but I haven't updated the benchmarks to use `testify` because the `require.NoError` check actually adds measurable overhead and screws with the measures (:sweat:).

Ready to merge. :ok_hand:

**Comment by deepthi** [comment]
> @harshit-gangal: I've added `testify` to all the new test cases, but I haven't updated the benchmarks to use `testify` because the `require.NoError` check actually adds measurable overhead and screws with the measures ().
> 
What about [quicktest](https://github.com/frankban/quicktest)? Does it add overhead?

**Comment by vmg** [comment]
@deepthi I haven't tested it, I just noticed the `testify` issue when re-running the benchmarks. I don't think it's particularly worrisome, Testify still works great for normal testing, so we can fall back to a simple `if err != nil` in the very few places where we have tight loops in benchmarks.

**Pull request merged**